File: java/sorting/BubbleSortProduction.java
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public final class BubbleSortProduction {

    private BubbleSortProduction() {
    }

    public static <T extends Comparable<? super T>> void bubbleSortInPlace(List<T> values) {
        bubbleSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void bubbleSortInPlace(List<T> values, Comparator<? super T> comparator) {
        Objects.requireNonNull(values, "values");
        Objects.requireNonNull(comparator, "comparator");

        int unsortedEnd = values.size() - 1;

        while (unsortedEnd > 0) {
            int lastSwap = 0;

            for (int index = 0; index < unsortedEnd; index++) {
                T left = values.get(index);
                T right = values.get(index + 1);

                if (comparator.compare(left, right) > 0) {
                    values.set(index, right);
                    values.set(index + 1, left);
                    lastSwap = index;
                }
            }

            if (lastSwap == 0) {
                break;
            }

            unsortedEnd = lastSwap;
        }
    }
}

File: c/sorting/bubble_sort_production.c
#include <stddef.h>

typedef int (*bubble_compare_fn)(const void *left, const void *right);

static void swap_bytes(unsigned char *left, unsigned char *right, size_t width)
{
    while (width-- > 0) {
        unsigned char temp = *left;
        *left++ = *right;
        *right++ = temp;
    }
}

int bubble_sort_in_place(void *base, size_t length, size_t width, bubble_compare_fn compare)
{
    if (base == NULL || compare == NULL || width == 0) {
        return -1;
    }

    if (length < 2) {
        return 0;
    }

    unsigned char *items = (unsigned char *)base;
    size_t unsorted_end = length - 1;

    while (unsorted_end > 0) {
        size_t last_swap = 0;

        for (size_t index = 0; index < unsorted_end; index++) {
            unsigned char *left = items + index * width;
            unsigned char *right = left + width;

            if (compare(left, right) > 0) {
                swap_bytes(left, right, width);
                last_swap = index;
            }
        }

        if (last_swap == 0) {
            break;
        }

        unsorted_end = last_swap;
    }

    return 0;
}
